File: 45_merge_intervals.js
/*
Given an array of intervals where intervals[i] = [starti, endi], 
merge all overlapping intervals, and return an array of the non-overlapping
intervals that cover all the intervals in the input.
*/

// https://leetcode.com/problems/merge-intervals/description/

/* sort & merge
sort the intervals by start time ASC
merge: extend the latest interval end if overlap
*/
var merge = function (intervals) {
  intervals.sort((a, b) => a[0] - b[0]);
  let result = [];
  for (let interval of intervals) {
    if (result.length === 0) result.push(interval);
    if (interval[0] > result[result.length - 1][1]) {
      result.push(interval);
    } else {
      result[result.length - 1][1] = Math.max(
        result[result.length - 1][1],
        interval[1]
      );
    }
  }
  return result;
};

File: 45_merge_intervals.py
# Given an array of intervals where intervals[i] = [starti, endi], 
# merge all overlapping intervals, and return an array of the non-overlapping
# intervals that cover all the intervals in the input.

# https://leetcode.com/problems/merge-intervals/
# related LC 57 https://leetcode.com/problems/insert-interval/

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()
        low, hi, l = intervals[0][0], intervals[0][1], len(intervals)
        ans = []
        for i in range(1, l):
            if hi < intervals[i][0]:
                ans.append([low, hi])
                low = intervals[i][0]
                hi = intervals[i][1]
            else:
                low = min(low, intervals[i][0])
                hi = max(hi, intervals[i][1])
        # append the last
        ans.append([low, hi])
        return ans
